File: lib/web_ui/lib/src/engine/compositor/picture.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// @dart = 2.10
part of engine;

class CkPicture implements ui.Picture {
  final SkiaObject<SkPicture> skiaObject;
  final ui.Rect? cullRect;

CkPicture(SkPicture picture, this.cullRect)
      : skiaObject = SkPictureSkiaObject(picture);

  @override
  int get approximateBytesUsed => 0;

  @override
  void dispose() {
    skiaObject.delete();
  }

  @override
  Future<ui.Image> toImage(int width, int height) async {
    final js.JsObject skSurface = canvasKit.callMethod('MakeSurface', <int>[width, height]);
    final js.JsObject skCanvas = skSurface.callMethod('getCanvas');
    skCanvas.callMethod('drawPicture', <js.JsObject>[skPicture.skiaObject!]);
    final js.JsObject skImage = skSurface.callMethod('makeImageSnapshot');
    skSurface.callMethod('dispose');
    return CkImage(skImage);
  }
}

class SkPictureSkiaObject extends OneShotSkiaObject<SkPicture> {
  SkPictureSkiaObject(SkPicture picture) : super(picture);

  @override
  void delete() {
    rawSkiaObject?.delete();
  }
}

File: lib/web_ui/skwasm/picture.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "export.h"
#include "helpers.h"
#include "render_strategy.h"
#include "wrappers.h"

using namespace Skwasm;

RTreeFactory bbhFactory;

SKWASM_EXPORT PictureRecorder* pictureRecorder_create() {
  return new PictureRecorder();
}

SKWASM_EXPORT void pictureRecorder_dispose(PictureRecorder* recorder) {
  delete recorder;
}

SKWASM_EXPORT Canvas* pictureRecorder_beginRecording(PictureRecorder* recorder,
                                                     const Rect* cullRect) {
  return recorder->beginRecording(*cullRect, &bbhFactory);
}

SKWASM_EXPORT Picture* pictureRecorder_endRecording(PictureRecorder* recorder) {
  return recorder->finishRecordingAsPicture().release();
}

SKWASM_EXPORT void picture_getCullRect(Picture* picture, Rect* outRect) {
  *outRect = picture->cullRect();
}

SKWASM_EXPORT void picture_dispose(Picture* picture) {
  picture->unref();
}

SKWASM_EXPORT uint32_t picture_approximateBytesUsed(Picture* picture) {
  return static_cast<uint32_t>(picture->approximateBytesUsed());
}
